packages/web/hooks/input/use-amount-input.ts

```diff
@@ -82,33 +104,37 @@ export function useAmountInput({
               .mul(DecUtils.getTenExponentN(currency.coinDecimals))
               .truncate();
 
-      const shouldSubtractMaxWithFee =
-        isMaxSelected &&
-        gasAmount?.denom === currency?.coinDenom &&
-        !!gasAmount;
-
       if (fraction != null && rawCurrencyBalance) {
         amountInt = new Dec(rawCurrencyBalance)
           .mul(new Dec(fraction))
-          .sub(
-            shouldSubtractMaxWithFee
-              ? new Dec(gasAmount.toCoin().amount)
-              : new Dec(0)
-          )
           .truncate();
       }
 
+      if (fraction === 1 && !isNil(maxAmount)) {
+        amountInt = new Int(maxAmount.toCoin().amount);
+      }
+
       if (amountInt.isZero()) return;
       return new CoinPretty(currency, amountInt);
     }
-  }, [
-    currency,
-    inputAmount,
-    isMaxSelected,
-    gasAmount,
-    fraction,
-    rawCurrencyBalance,
-  ]);
+  }, [currency, inputAmount, fraction, rawCurrencyBalance, maxAmount]);
+
+  const isHalfValue = useMemo(
+    () =>
+      fraction === 0.5 ||
+      (rawCurrencyBalance && amount
+        ? new Dec(amount.toCoin().amount).equals(
+            new Dec(rawCurrencyBalance).mul(new Dec(0.5))
+          )
+        : false),
+    [amount, fraction, rawCurrencyBalance]
+  );
+  const isMaxValue = useMemo(
+    () =>
+      fraction === 1 ||
+      (amount && maxAmount ? amount.toDec().equals(maxAmount.toDec()) : false),
```

Comment:
What about a value that is not equal to max, but is still high enough to cause the out of balance gas error?

Reply:
This error is handled on [these](https://github.com/osmosis-labs/osmosis-frontend/pull/3132/files/f7b0c7533183a6333a78cd46c4189259c9664b8a#diff-c4bd127a6249c1fc41dc3ea0d6f89afb55d71b6bfd22f9e731e94c7cd1804c76R189-R191) lines. There is also a [test](https://github.com/osmosis-labs/osmosis-frontend/pull/3132/files/f7b0c7533183a6333a78cd46c4189259c9664b8a#diff-69ecd68b40ff73e084ebec4e17d6d2ee010c11d3535936e5a4b161abee678e72R153-R164)

Reply:
Hmm. Like consider the scenario where a total tx amount is 10 OSMO, gas is 1 OSMO, and the user manually enters 9.1 OSMO as the amount.

They would be short 0.1 OSMO to pay the fee, and I'm not sure this logic would handle that unless the user entered 10 OSMO (either with max or manually)

Reply:
Perhaps I'm overlooking something, but we would encounter an insufficient balance error if the user doesn't have enough funds for this transaction. Are you suggesting that there should be a different response or behavior?

Reply:
Oh ok. I may not have looked at the logic close enough. Thanks for confirming.